packages/core/src/rules/oas2/remove-unused-components.ts

```diff
@@ -0,0 +1,88 @@
+import { Oas2Rule } from '../../visitors';
+import { Location, parsePointer } from '../../ref-utils';
+import { Oas2Components } from '../../typings/swagger';
+import { isEmptyObject } from '../../utils';
+
+export const RemoveUnusedComponents: Oas2Rule = () => {
+  let components = new Map<string, { used: boolean; componentType?: keyof Oas2Components; name: string }>();
+
+  function registerComponent(location: Location, componentType: keyof Oas2Components, name: string): void {
+    components.set(location.absolutePointer, {
+      used: components.get(location.absolutePointer)?.used || false,
+      componentType,
+      name,
+    });
+  }
+
+  return {
+    ref(ref, { type, resolve, key }) {
+      if (
+        ['Schema', 'Parameter', 'Response', 'SecurityScheme'].includes(type.name)
+      ) {
+        const resolvedRef = resolve(ref);
+        if (!resolvedRef.location) return;
+        components.set(resolvedRef.location.absolutePointer, {
+          used: true,
+          name: key.toString(),
+        });
+      }
+    },
+
+    DefinitionRoot: {
+      leave(root) {
+        let rootComponents = new Set<keyof Oas2Components>();
+        components.forEach(usageInfo => {
+          const { used, name, componentType } = usageInfo;
+          if (!used && componentType) {
+            rootComponents.add(componentType);
+            delete root[componentType]![name];
+          }
+        });
+        for (const component of rootComponents) {
+          if (isEmptyObject(root[component])) {
+            delete root[component];
+          }
+        }
+      },
+    },
+
+    NamedSchemas: {
+      Schema(schema, { location, key }) {
+        if (!schema.allOf) {
+          registerComponent(
+            location,
+            parsePointer(location.pointer)[0] as keyof Oas2Components,
```

Comment:
This is not correct in all the cases, I already mentioned it before but it was not resolved: https://github.com/Redocly/openapi-cli/pull/443#discussion_r776966814)

> Do not parse pointer here, it is error prone, pointer can be to external file, instead just hardcode component type here

Reply:
done